wallet/src/events.rs

```diff
@@ -1,15 +1,97 @@
 use crate::address::Role;
+use ic_cdk::export::candid::types::{Compound, Serializer, Type};
 use ic_cdk::export::candid::CandidType;
 use ic_cdk::export::Principal;
 use ic_cdk::{api, storage};
-use serde::Deserialize;
+use indexmap::IndexMap;
+use serde::de::{SeqAccess, Visitor};
+use serde::{Deserialize, Deserializer};
 use std::cmp::min;
+use std::fmt::{self, Formatter};
 
 #[derive(CandidType, Clone, Default, Deserialize)]
 pub struct EventBuffer {
     events: Vec<Event>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct ManagedList(IndexMap<Principal, ManagedCanister>);
+
+impl ManagedList {
+    pub fn push(&mut self, canister: Principal, event: ManagedCanisterEventKind) {
+        self.push_with_timestamp(canister, event, api::time())
+    }
+    pub fn push_with_timestamp(
+        &mut self,
+        canister: Principal,
+        event: ManagedCanisterEventKind,
+        timestamp: u64,
+    ) {
+        let events = &mut self
+            .0
+            .entry(canister)
+            .or_insert_with(|| ManagedCanister::new(canister))
+            .events;
+        events.push(ManagedCanisterEvent {
+            kind: event,
+            id: events.len() as u32,
+            timestamp,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Eq, CandidType, Deserialize)]
+pub struct ManagedCanister {
+    pub info: ManagedCanisterInfo,
+    pub events: Vec<ManagedCanisterEvent>,
+}
+
+#[derive(Debug, Clone, Eq, CandidType, Deserialize)]
+pub struct ManagedCanisterInfo {
+    pub id: Principal,
+    pub name: Option<String>,
+    pub created_at: i64,
+}
+
+impl ManagedCanister {
+    pub fn new(id: Principal) -> Self {
+        Self {
+            info: ManagedCanisterInfo {
+                id,
+                name: None,
+                created_at: api::time() as i64,
+            },
+            events: vec![],
+        }
+    }
+}
+
+impl PartialEq for ManagedCanisterInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl PartialEq for ManagedCanister {
+    fn eq(&self, other: &Self) -> bool {
+        self.info == other.info
```

Comment:
Would it make sense for this to compare `self.info.id == other.info.id` instead? Is there a reason to include `name` and `created_at` in partial equality here?

Reply:
It doesn't include `name` and `created_at`, those are left out in `<ManagedCanisterInfo as PartialEq>`. This just makes sure that you only have to change that behavior in one place.